src/relay/index.ts

```diff
@@ -0,0 +1,49 @@
+import { multiaddr } from '@multiformats/multiaddr'
+import { expect } from 'aegir/chai'
+import type { Daemon, DaemonFactory, NodeType, SpawnOptions } from '../index.js'
+import { Status } from './pb/index.js'
+import { reserve } from './util.js'
+
+export function relayTests (factory: DaemonFactory) {
+  const t: NodeType[] = ['go', 'js']
+  t.forEach(a => t.forEach(b => t.forEach(r => { if (a !== b || a !== r) relayTest(factory, a, b, r) })))
+}
+
+function relayTest (factory: DaemonFactory, aType: NodeType, bType: NodeType, relayType: NodeType) {
+  describe(`${aType} to ${bType} over relay ${relayType}`, () => {
+    let daemons: Daemon[] = []
+    const opts: SpawnOptions[] = [
+      { type: aType, noise: true, noListen: true },
+      { type: bType, noise: true, noListen: true },
+      { type: relayType, noise: true, relay: true }
+    ]
+
+    beforeEach(async function () {
+      this.timeout(20 * 1000)
+      daemons = await Promise.all(opts.map(async o => await factory.spawn(o)))
+    })
+
+    afterEach(async function () {
+      await Promise.all(daemons.map(async d => await d.stop()))
+    })
+
+    it('connects', async () => {
+      const [aNode, bNode] = daemons
+      const identify = await Promise.all(daemons.map(async d => await d.client.identify()))
+      const bId = identify[1]
+      const relayId = identify[2]
+
+      // connect receiver to the relay
+      await bNode.client.connect(relayId.peerId, relayId.addrs)
+      const reserveResponse = await reserve(bNode, relayId.peerId)
+      expect(reserveResponse.status).to.eq(Status.OK)
+
+      // construct a relay address
+      const addr = multiaddr(`${relayId.addrs[0].toString()}/p2p/${relayId.peerId.toString()}/p2p-circuit/p2p/${bId.peerId.toString()}`)
+      await aNode.client.connect(bId.peerId, [addr])
+      await new Promise(resolve => setTimeout(resolve, 500))
+      const connectedPeers = await aNode.client.listPeers()
+      expect(connectedPeers.filter(p => p.equals(bId.peerId))).to.have.length(1)
```

Comment:
so you know you test the communication channel.

Reply:
Yes, I'll do that

Reply:
@MarcoPolo While writing an echo test (daemon does not have ping), I found that go-libp2p marks the relayed connections as transient. This leads to an issue where the daemon is unable to open streams since it does not have `WithUseTransient` enabled on stream creation. https://github.com/libp2p/go-libp2p-daemon/blob/master/conn.go#L228-L243
While this is easily fixable, it does require an additional change to the go-libp2p-daemon.